odl/contrib/solvers/spdhg/misc.py

```diff
@@ -0,0 +1,857 @@
+
+    with regularization parameter alpha and strong convexity beta. In addition,
+    the nonnegativity constraint is achieved with the characteristic function
+
+        char_fun(x) = 0 if x >= 0 and infty else.
+
+    Parameters
+    ----------
+    domain : odlspace
+        domain of TV functional
+    alpha : scalar, optional
+        Regularization parameter, positive
+    prox_options : dict, optional
+        name: string, optional
+            name of the method to perform the prox operator, default=FGP
+        warmstart: boolean, optional
+            Do you want a warm start, i.e. start with the dual variable
+            from the last call? default=True
+        niter: int, optional
+            number of iterations per call, default=5
+        p: array, optional
+            initial dual variable, default=zeros
+    grad : gradient operator, optional
+        Gradient operator to be used within the total variation functional.
+        default=see TV
+    """
+
+    def __init__(self, domain, alpha=1, prox_options={}, grad=None,
+                 strong_convexity=0):
+        """
+        """
+
+        self.strong_convexity = strong_convexity
+
+        if 'name' not in prox_options:
+            prox_options['name'] = 'FGP'
+        if 'warmstart' not in prox_options:
+            prox_options['warmstart'] = True
+        if 'niter' not in prox_options:
+            prox_options['niter'] = 5
+        if 'p' not in prox_options:
+            prox_options['p'] = None
+        if 'tol' not in prox_options:
+            prox_options['tol'] = None
+
+        self.prox_options = prox_options
+
+        self.alpha = alpha
+        self.TV = TV(domain, grad=grad)
+        self.grad = self.TV.right
+        self.NN = odl.solvers.IndicatorBox(domain, 0, np.inf)
+        self.L2 = 0.5 * odl.solvers.L2NormSquared(domain)
+
+        super().__init__(space=domain, linear=False, grad_lipschitz=0)
+
+    def __call__(self, x):
+        """ Characteristic function of the non-negative orthant
+
+        Parameters
+        ----------
+        x : np.array
+            vector / image
+
+        Returns
+        -------
+        extended float (with infinity)
+            Is the input in the non-negative orthant?
+
+        Examples
+        --------
+        Check that the total variation of a constant is zero
+
+        >>> import odl.contrib.spdhg as spdhg, odl
+        >>> space = odl.uniform_discr([0, 0], [3, 3], [3, 3])
+        >>> TVNN = spdhg.TV_NonNegative(space, alpha=2)
+        >>> x = space.one()
+        >>> TVNN(x) < 1e-10
+
+        Check that negative functions are mapped to infty
+
+        >>> import odl.contrib.spdhg as spdhg, odl, numpy as np
+        >>> space = odl.uniform_discr([0, 0], [3, 3], [3, 3])
+        >>> TVNN = spdhg.TV_NonNegative(space, alpha=2)
+        >>> x = -space.one()
+        >>> np.isinf(TVNN(x))
+        """
+
+        nn = self.NN(x)
+
+        if nn is np.inf:
```

Comment:
This is extremely fragile and breaks as soon as we apply `float()` to the output of `IndicatorBox`. Better compare with `==`, and why not use core Python `float('inf')`?